C error only with `-autofree` flag.

### Describe the bug

A C error is generated from `maybe_error()!` only when `-autofree` flag is set.

### Reproduction Steps

```v
import os

fn maybe_error() !string {
	return 'string'
}

fn return_string_result() ! {
	os.write_file('./test.txt', maybe_error()!)!
}

fn main() {
	return_string_result() or {
		println(err.msg())
		exit(1)
	}
}
```

### Expected Behavior

Expected code to write a file with `string` inside of it.

### Current Behavior

```
==================
C:/Users/imado/AppData/Local/Temp/v_0/test.7251254077606544120.tmp.c:18212: warning: assignment from incompatible pointer type
C:/Users/imado/AppData/Local/Temp/v_0/test.7251254077606544120.tmp.c:18307: error: 'esult_string' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V full version: V 0.4.2 ae5b4bb.c030c5e

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.2 ae5b4bb.c030c5e
OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\Users\imado\Documents\GitHub\time_without
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-09-30 16:48:31

vroot: OK, value: C:\Users\imado\v
VMODULES: OK, value: C:\Users\imado\.vmodules
VTMP: OK, value: C:\Users\imado\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: 0.4.2-3-gc030c5ef (14 commit(s) behind V master)
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -531,7 +531,8 @@ fn (mut c Checker) call_expr(mut node ast.CallExpr) ast.Type {
 			if arg.typ != ast.string_type {
 				continue
 			}
-			if arg.expr in [ast.Ident, ast.StringLiteral, ast.SelectorExpr] {
+			if arg.expr in [ast.Ident, ast.StringLiteral, ast.SelectorExpr]
+				|| (arg.expr is ast.CallExpr && arg.expr.or_block.kind != .absent) {
 				// Simple expressions like variables, string literals, selector expressions
 				// (`x.field`) can't result in allocations and don't need to be assigned to
 				// temporary vars.

